### PR TITLE
Handle key collision with duplicate definitions using same parameters

### PR DESCRIPTION
**Pull request by ArmaanMcleod**
## PR Summary

Related to #1653.

Fixes key collisions with definitions using same parameters.
Also refactored a bit of csharp code.

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [ ] Unit tests created/ updated
  - [ ] Rule documentation created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [x] Unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs [hunk 1]
```diff
@@ -673,6 +673,27 @@ protected virtual void Definitions(PolicyAssignmentContext context, IEnumerable<
...
+            if (typeA == ParameterType.String && typeB == ParameterType.String)
+                return valueA.ToString().Equals(valueB.ToString(), StringComparison.OrdinalIgnoreCase);
+
+            if (typeA == ParameterType.Integer && typeB == ParameterType.Integer)
+                return Convert.ToInt32(valueA, Thread.CurrentThread.CurrentCulture) == Convert.ToInt32(valueB, Thread.CurrentThread.CurrentCulture);
+
+            if (typeA == ParameterType.Boolean && typeB == ParameterType.Boolean)
+                return Convert.ToBoolean(valueA, Thread.CurrentThread.CurrentCulture) == Convert.ToBoolean(valueB, Thread.CurrentThread.CurrentCulture);
+
+            // TODO: Handle more types
```
Comment (ArmaanMcleod) [inline src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs:692]: I didn't include all types here. Can handle the more tricky ones if needed.

File: src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs [hunk 2]
```diff
@@ -710,7 +734,36 @@ protected virtual bool TryPolicyDefinition(PolicyAssignmentContext context, JObj
...
+                        context.DefinitionIds[policyDefinitionId] = result;
+
+                    // Otherwise skip adding definition
+                    else
+                        return false;
+                }
+
+                // Add definition for the first time
+                else
+                    context.DefinitionIds.Add(policyDefinitionId, result);
```
Comment (BernieWhite) [inline src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs:766]: @ArmaanMcleod Do we even need to store the policy definitions in `context.DefinitionIds` as a dictionary? Couldn't we just care about the current policy definition we are processing only.

The only real reference to this is: https://vscode.dev/github/armaanmcleod/psrule.rules.azure/blob/607b51868b45767215d852c2bb1644151e2faf5c/src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs#L584-L589

After we've processed a policy definition we move on to the next one. Sure, if we reprocess the same definition there might be a little more processing for same policy + same parameters. But storing all policies in memory, also has a hit.

Comment (ArmaanMcleod) [reply src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs:766]: @BernieWhite I think we could just store the definition parameters, that would probably reduce memory footprint alot.

I probably left it like this when I was getting inspiration from the `TemplateVisitor` code 😄.

**Comment by ArmaanMcleod** [comment]
@BernieWhite Made some changes to only store definition parameters, let me know what you think.

File: tests/PSRule.Rules.Azure.Tests/test11.assignment.json [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    },
+    "PolicyDefinitions": [
+      {
+        "Name": "00000000-0000-0000-0000-000000000000",
+        "ResourceId": "/providers/Microsoft.Authorization/policyDefinitions/00000000-0000-0000-0000-000000000000",
+        "ResourceName": "00000000-0000-0000-0000-000000000000",
+        "ResourceType": "Microsoft.Authorization/policyDefinitions",
+        "SubscriptionId": null,
+        "Properties": {
+          "Description": "CMA_0544 - View and configure system diagnostic data",
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/test11.assignment.json:39]:
```suggestion
          "Description": "View and configure system diagnostic data",
```

File: tests/PSRule.Rules.Azure.Tests/test11.assignment.json [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        "PolicyDefinitionId": "/providers/Microsoft.Authorization/policyDefinitions/00000000-0000-0000-0000-000000000000"
+      },
+      {
+        "Name": "00000000-0000-0000-0000-000000000000",
+        "ResourceId": "/providers/Microsoft.Authorization/policyDefinitions/00000000-0000-0000-0000-000000000000",
+        "ResourceName": "00000000-0000-0000-0000-000000000000",
+        "ResourceType": "Microsoft.Authorization/policyDefinitions",
+        "SubscriptionId": null,
+        "Properties": {
+          "Description": "CMA_0544 - View and configure system diagnostic data",
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/test11.assignment.json:84]:
```suggestion
          "Description": "View and configure system diagnostic data",
```

File: tests/PSRule.Rules.Azure.Tests/test11.assignment.json [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+          "DisplayName": "View and configure system diagnostic data",
+          "Metadata": {
+            "version": "1.0.0",
+            "category": "Regulatory Compliance",
+            "additionalMetadataId": "/providers/Microsoft.PolicyInsights/policyMetadata/CMA_0544"
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/test11.assignment.json:44]:
```suggestion
            "category": "Regulatory Compliance"
```

File: tests/PSRule.Rules.Azure.Tests/test11.assignment.json [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/test11.assignment.json:89]:
```suggestion
            "category": "Regulatory Compliance"
```

File: tests/PSRule.Rules.Azure.Tests/emittedJsonRulesData.jsonc [hunk 7]
```diff
@@ -641,5 +641,29 @@
         ]
       }
     }
+  },
+  {
+    // Synopsis: CMA_0544 - View and configure system diagnostic data
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/emittedJsonRulesData.jsonc:646]:
```suggestion
    // Synopsis: View and configure system diagnostic data
```

**Review by BernieWhite — CHANGES_REQUESTED** [review]
@ArmaanMcleod Nice work. I think functionally let's go with this. I think we can optimize this further over the next few releases, however I don't think we need to invest that time now as the code based is rapidly changing.

Let's fix these id's and after that we should be good to merge.

**Review by BernieWhite — APPROVED** [review]
Thanks @ArmaanMcleod all good to merge.

**Pull request merged**